pkg/apis/machine/v1beta1/machineset_webhook_test.go

```diff
@@ -0,0 +1,600 @@
+		return cfg, nil
+	}
+	defer func() {
+		ctrl.GetConfig = config.GetConfig
+	}()
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "machineset-creation-test",
+		},
+	}
+	g.Expect(c.Create(ctx, namespace)).To(Succeed())
+	defer func() {
+		g.Expect(c.Delete(ctx, namespace)).To(Succeed())
+	}()
+
+	infra := &osconfigv1.Infrastructure{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cluster",
+		},
+	}
+	g.Expect(c.Create(ctx, infra)).To(Succeed())
+	defer func() {
+		g.Expect(c.Delete(ctx, infra)).To(Succeed())
+	}()
+
+	testCases := []struct {
+		name              string
+		platformType      osconfigv1.PlatformType
+		clusterID         string
+		expectedError     string
+		providerSpecValue *runtime.RawExtension
+	}{
+		{
+			name:         "with AWS and no fields set",
+			platformType: osconfigv1.AWSPlatformType,
+			clusterID:    "aws-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &aws.AWSMachineProviderConfig{},
+			},
+			expectedError: "providerSpec.ami: Required value: expected either providerSpec.ami.arn or providerSpec.ami.filters or providerSpec.ami.id to be populated",
+		},
+		{
+			name:         "with AWS and an AMI ID set",
+			platformType: osconfigv1.AWSPlatformType,
+			clusterID:    "aws-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &aws.AWSMachineProviderConfig{
+					AMI: aws.AWSResourceReference{
+						ID: pointer.StringPtr("ami"),
+					},
+				},
+			},
+			expectedError: "",
+		},
+		{
+			name:         "with Azure and no fields set",
+			platformType: osconfigv1.AzurePlatformType,
+			clusterID:    "azure-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &azure.AzureMachineProviderSpec{},
+			},
+			expectedError: "[providerSpec.location: Required value: location should be set to one of the supported Azure regions, providerSpec.osDisk.diskSizeGB: Invalid value: 0: diskSizeGB must be greater than zero]",
+		},
+		{
+			name:         "with Azure and a location and disk size set",
+			platformType: osconfigv1.AzurePlatformType,
+			clusterID:    "azure-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &azure.AzureMachineProviderSpec{
+					Location: "location",
+					OSDisk: azure.OSDisk{
+						DiskSizeGB: 128,
+					},
+				},
+			},
+			expectedError: "",
+		},
+		{
+			name:         "with GCP and no fields set",
+			platformType: osconfigv1.GCPPlatformType,
+			clusterID:    "gcp-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &gcp.GCPMachineProviderSpec{},
+			},
+			expectedError: "providerSpec.region: Required value: region is required",
+		},
+		{
+			name:         "with GCP and the region and zone set",
+			platformType: osconfigv1.GCPPlatformType,
+			clusterID:    "gcp-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &gcp.GCPMachineProviderSpec{
+					Region: "region",
+					Zone:   "region-zone",
+				},
+			},
+			expectedError: "",
+		},
```

Comment:
ProviderSpec.Value can be nil.

```
		{
			name:         "with GCP and nil providerSpec",
			platformType: osconfigv1.GCPPlatformType,
			clusterID:    "gcp-cluster",
			providerSpecValue: nil,
			expectedError: "x",
		},
```

Reply:
Good catch, added some tests and a check that ensures the provider spec value isn't nil